File: src/algorithms/growth.jl
"""
Growth and pruning functions for BasicDBGSOM.
"""

# ============================================================================
# Error Accumulation
# ============================================================================

"""
    reset_errors!(topo)

Reset accumulated errors and hit counts for all neurons.
"""
function reset_errors!(topo::SOMTopology{T}) where T
    for neuron in values(topo.neurons)
        neuron.error = zero(T)
        neuron.hits = 0
    end
end

reset_errors!(som::DBGSOM) = reset_errors!(som.topology)

"""
    accumulate_errors!(topo, X, bmus)

Accumulate quantization errors for each neuron based on BMU assignments.
Handles NaN values by using scaled partial distances.

# Arguments
- `topo::SOMTopology{T}`: Topology to update
- `X::AbstractMatrix{T}`: Input data (input_dim × n_samples)
- `bmus::Vector{Tuple{Int,Int}}`: BMU positions for each sample
"""
function accumulate_errors!(topo::SOMTopology{T}, X::AbstractMatrix{T},
                            bmus::Vector{Tuple{Int,Int}}) where T
    n_samples = size(X, 2)

    @inbounds for i in 1:n_samples
        pos = bmus[i]
        neuron = topo.neurons[pos]
        x = view(X, :, i)

        # Compute distance to sample (NaN-aware)
        dist = if any(isnan, x)
            euclidean_nan(neuron.weights, x)
        else
            euclidean(neuron.weights, x)
        end

        # Only accumulate if distance is valid
        if !isnan(dist) && dist < typemax(T)
            neuron.error += dist
            neuron.hits += 1
        end
    end
end

accumulate_errors!(som::DBGSOM{T}, X::AbstractMatrix{T},
                   bmus::Vector{Tuple{Int,Int}}) where T =
    accumulate_errors!(som.topology, X, bmus)

"""
    accumulate_errors!(topo, X, bmus, miss_info)

Accumulate quantization errors using missingness-weighted distances when
`miss_info` is provided. Falls back to standard distances when `miss_info`
is `nothing`.
"""
function accumulate_errors!(topo::SOMTopology{T}, X::AbstractMatrix{T},
                            bmus::Vector{Tuple{Int,Int}},
                            miss_info::Union{Nothing, MissingnessInfo{T}}) where T
    if miss_info === nothing
        return accumulate_errors!(topo, X, bmus)
    end

    n_samples = size(X, 2)

    @inbounds for i in 1:n_samples
        pos = bmus[i]
        neuron = topo.neurons[pos]
        x = view(X, :, i)

        # Compute distance using missingness-weighted metric
        dist_sq = if any(isnan, x)
            euclidean_squared_nan_weighted(neuron.weights, x, miss_info)
        else
            euclidean_squared(neuron.weights, x)
        end

        dist = sqrt(dist_sq)

        # Only accumulate if distance is valid
        if !isnan(dist) && dist < typemax(T)
            neuron.error += dist
            neuron.hits += 1
        end
    end
end

accumulate_errors!(som::DBGSOM{T}, X::AbstractMatrix{T},
                   bmus::Vector{Tuple{Int,Int}},
                   miss_info::Union{Nothing, MissingnessInfo{T}}) where T =
    accumulate_errors!(som.topology, X, bmus, miss_info)

# ============================================================================
# Error Distribution
# ============================================================================

"""
    distribute_errors!(topo, gt)

Distribute errors from non-boundary neurons to their boundary neighbors.

For each non-boundary neuron where E_i ≥ GT:
- Each boundary neighbor receives error/12 (hexagonal grid has up to 12 second-order neighbors)
- The non-boundary neuron retains error/2

Adapted from Vasighi & Amini (2017) for hexagonal grid topology.
"""
function distribute_errors!(topo::SOMTopology{T}, gt::T) where T
    # Collect non-boundary neurons with high error
    for (pos, neuron) in topo.neurons
        if neuron.error >= gt && !is_boundary(topo, pos)
            # Find boundary neighbors (direct neighbors that are on boundary)
            boundary_neighbors = Tuple{Int,Int}[]
            for neighbor_pos in get_neighbors(pos)
                if has_neuron(topo, neighbor_pos) && is_boundary(topo, neighbor_pos)
                    push!(boundary_neighbors, neighbor_pos)
                end
            end

            # Distribute error/12 to each boundary neighbor, retain error/2
            # (12 accounts for hexagonal second-order neighborhood size)
            if !isempty(boundary_neighbors)
                error_share = neuron.error / T(12)
                for neighbor_pos in boundary_neighbors
                    topo.neurons[neighbor_pos].error += error_share
                end
                neuron.error = neuron.error / T(2)
            end
        end
    end
end

distribute_errors!(som::DBGSOM{T}, gt::T) where T =
    distribute_errors!(som.topology, gt)

# ============================================================================
# Growth Candidates
# ============================================================================

"""
    find_growth_candidates(topo, gt) -> Vector{Tuple{Int,Int}}

Find boundary neurons with error exceeding growth threshold.

# Arguments
- `topo::SOMTopology{T}`: Topology to search
- `gt::T`: Growth threshold

# Returns
Positions sorted by error (highest first)
"""
function find_growth_candidates(topo::SOMTopology{T}, gt::T) where T
    candidates = Tuple{Int,Int}[]

    for (pos, neuron) in topo.neurons
        if neuron.error > gt && is_boundary(topo, pos)
            push!(candidates, pos)
        end
    end

    # Sort by error (highest first)
    sort!(candidates, by=pos -> topo.neurons[pos].error, rev=true)

    return candidates
end

find_growth_candidates(som::DBGSOM{T}, gt::T) where T =
    find_growth_candidates(som.topology, gt)

# ============================================================================
# Directed Growth (Vasighi & Amini 2017)
# ============================================================================

"""
    select_growth_direction(topo, boundary_pos) -> (new_pos, weights) or nothing

Select the best growth direction from a boundary neuron based on error distribution
of neighbors. This implements the "directed" part of DBGSOM for hexagonal grids.

The algorithm considers:
- 1st-order neighbors (distance=1): direct neighbors (up to 6 in hex grid)
- 2nd-order neighbors (distance=2): neighbors of neighbors
- Growth direction is biased toward regions with higher accumulated error

Adapted from Vasighi & Amini (2017) for hexagonal grid topology.

# Returns
Tuple of (new_position, weight_vector) or nothing if no growth possible
"""
function select_growth_direction(topo::SOMTopology{T}, boundary_pos::Tuple{Int,Int}) where T
    parent = topo.neurons[boundary_pos]
    allowed = get_empty_neighbors(topo, boundary_pos)

    if isempty(allowed)
        return nothing
    end

    # Find 1st-order neighbors (distance=1) and 2nd-order neighbors (distance=2)
    nbr1 = get_neighbors_at_distance(topo, boundary_pos, 1)  # 0-5 occupied (out of 6)
    nbr2 = get_neighbors_at_distance(topo, boundary_pos, 2)  # 2nd order neighbors

    # Sort neighbors by error (highest first)
    sort!(nbr1, by=p -> topo.neurons[p].error, rev=true)
    sort!(nbr2, by=p -> topo.neurons[p].error, rev=true)

    n_nbr1 = length(nbr1)

    # Hex cases: boundary neuron has 0-5 occupied neighbors (6 - n_allowed)
    if n_nbr1 == 0
        # Isolated neuron - grow in random direction
        return _grow_isolated_hex(topo, boundary_pos, allowed)
    elseif n_nbr1 <= 2
        # Few neighbors (1-2): grow away from neighbors toward high-error region
        return _grow_few_neighbors_hex(topo, boundary_pos, allowed, nbr1, nbr2)
    elseif n_nbr1 <= 4
        # Medium neighbors (3-4): grow toward highest error direction
        return _grow_medium_neighbors_hex(topo, boundary_pos, allowed, nbr1, nbr2)
    else
        # Many neighbors (5): only one slot available
        return _grow_many_neighbors_hex(topo, boundary_pos, allowed, nbr1, nbr2)
    end
end

"""
    _grow_isolated_hex(topo, bo_pos, allowed)

Growth when boundary neuron has no neighbors (isolated).
Copies parent weights with small noise.
"""
function _grow_isolated_hex(topo::SOMTopology{T}, bo_pos::Tuple{Int,Int},
                            allowed::Vector{Tuple{Int,Int}}) where T
    parent = topo.neurons[bo_pos]
    new_pos = allowed[1]
    weights = copy(parent.weights)
    weights .+= randn(T, length(weights)) .* T(0.01)
    return (new_pos, weights)
end

"""
    _grow_few_neighbors_hex(topo, bo_pos, allowed, nbr1, nbr2)

Growth when boundary neuron has 1-2 direct neighbors (4-5 empty slots).
Prefers growing toward highest-error 2nd-order neighbor, or away from existing neighbors.
"""
function _grow_few_neighbors_hex(topo::SOMTopology{T}, bo_pos::Tuple{Int,Int},
                                  allowed::Vector{Tuple{Int,Int}},
                                  nbr1::Vector{Tuple{Int,Int}},
                                  nbr2::Vector{Tuple{Int,Int}}) where T
    parent = topo.neurons[bo_pos]

    # Find allowed positions adjacent to high-error nbr2
    sel_nbr2 = filter(p -> any(grid_distance(p, a) == 1 for a in allowed), nbr2)

    if !isempty(sel_nbr2)
        sort!(sel_nbr2, by=p -> topo.neurons[p].error, rev=true)
        best_nbr2 = topo.neurons[sel_nbr2[1]]
        best_nbr2_pos = sel_nbr2[1]
        # Find allowed position closest to best_nbr2
        new_pos = argmin(a -> grid_distance(a, best_nbr2_pos), allowed)
        weights = (parent.weights .+ best_nbr2.weights) ./ T(2)
    else
        # Grow away from first neighbor using reflection
        nbr = topo.neurons[nbr1[1]]
        nbr1_pos = nbr1[1]
        # Pick position farthest from nbr1[1]
        new_pos = argmax(a -> grid_distance(a, nbr1_pos), allowed)
        weights = 2 .* parent.weights .- nbr.weights
    end

    return (new_pos, weights)
end

"""
    _grow_medium_neighbors_hex(topo, bo_pos, allowed, nbr1, nbr2)

Growth when boundary neuron has 3-4 direct neighbors (2-3 empty slots).
Grows toward the direction with highest accumulated error.
"""
function _grow_medium_neighbors_hex(topo::SOMTopology{T}, bo_pos::Tuple{Int,Int},
                                     allowed::Vector{Tuple{Int,Int}},
                                     nbr1::Vector{Tuple{Int,Int}},
                                     nbr2::Vector{Tuple{Int,Int}}) where T
    parent = topo.neurons[bo_pos]
    max_err_nbr_pos = nbr1[1]  # Highest error neighbor (sorted)
    min_err_nbr = topo.neurons[nbr1[end]]  # Lowest error neighbor

    # Pick allowed position closest to highest-error neighbor
    new_pos = argmin(a -> grid_distance(a, max_err_nbr_pos), allowed)

    # Use reflection from lowest-error neighbor
    weights = 2 .* parent.weights .- min_err_nbr.weights

    return (new_pos, weights)
end

"""
    _grow_many_neighbors_hex(topo, bo_pos, allowed, nbr1, nbr2)

Growth when boundary neuron has 5 direct neighbors (1 empty slot).
Only one direction possible, uses opposite neighbor for reflection.
"""
function _grow_many_neighbors_hex(topo::SOMTopology{T}, bo_pos::Tuple{Int,Int},
                                   allowed::Vector{Tuple{Int,Int}},
                                   nbr1::Vector{Tuple{Int,Int}},
                                   nbr2::Vector{Tuple{Int,Int}}) where T
    parent = topo.neurons[bo_pos]
    new_pos = allowed[1]

    # Find opposite neighbor (distance 2 from new_pos)
    opposite = filter(n -> grid_distance(n, new_pos) == 2, nbr1)

    if !isempty(opposite)
        opp_nbr = topo.neurons[opposite[1]]
        weights = 2 .* parent.weights .- opp_nbr.weights
    else
        # No clear opposite - average reflection from all neighbors
        avg_weights = mean([topo.neurons[n].weights for n in nbr1])
        weights = 2 .* parent.weights .- avg_weights
    end

    return (new_pos, weights)
end

"""
    insert_neuron_directed!(topo, pos, weights)

Insert a new neuron with pre-computed weights at the specified position.
"""
function insert_neuron_directed!(topo::SOMTopology{T}, pos::Tuple{Int,Int},
                                  weights::Vector{T}) where T
    @assert !has_neuron(topo, pos) "Position $pos already occupied"
    topo.neurons[pos] = Neuron{T}(weights)
end

insert_neuron_directed!(som::DBGSOM{T}, pos::Tuple{Int,Int}, weights::Vector{T}) where T =
    insert_neuron_directed!(som.topology, pos, weights)

# ============================================================================
# Legacy Neuron Insertion (kept for compatibility)
# ============================================================================

"""
    insert_neuron!(topo, pos, parent_pos)

Insert a new neuron at pos, initialized from parent neuron using reflection formula.
This is the simpler version; prefer using select_growth_direction + insert_neuron_directed!
for the full directed growth algorithm.
"""
function insert_neuron!(topo::SOMTopology{T}, pos::Tuple{Int,Int},
                        parent_pos::Tuple{Int,Int}) where T
    @assert !has_neuron(topo, pos) "Position $pos already occupied"
    @assert has_neuron(topo, parent_pos) "Parent position $parent_pos not found"

    parent = topo.neurons[parent_pos]
    occupied_neighbors = get_occupied_neighbors(topo, parent_pos)

    if !isempty(occupied_neighbors)
        neighbor_pos = occupied_neighbors[1]
        neighbor = topo.neurons[neighbor_pos]
        weights = 2 .* parent.weights .- neighbor.weights
    else
        weights = copy(parent.weights)
        weights .+= randn(T, length(weights)) .* T(0.01)
    end

    topo.neurons[pos] = Neuron{T}(weights)
end

insert_neuron!(som::DBGSOM{T}, pos::Tuple{Int,Int}, parent_pos::Tuple{Int,Int}) where T =
    insert_neuron!(som.topology, pos, parent_pos)

# ============================================================================
# Growth Step
# ============================================================================

"""
    grow!(som, gt) -> Int

Grow neurons from high-error boundary neurons using directed growth.

Uses the full DBGSOM directed growth algorithm:
- Considers 1st and 2nd-order neighbors
- Growth direction biased toward high-error regions
- Weight initialization uses reflection with neighbor error weighting

Reference: Vasighi & Amini (2017)

# Arguments
- `som::DBGSOM{T}`: SOM to grow
- `gt::T`: Growth threshold

# Returns
Number of neurons added
"""
function grow!(som::DBGSOM{T}, gt::T) where T
    candidates = find_growth_candidates(som, gt)
    n_added = 0

    for pos in candidates
        if n_neurons(som) >= som.max_neurons
            break
        end

        # Use directed growth to select position and compute weights
        result = select_growth_direction(som.topology, pos)

        if result !== nothing
            new_pos, weights = result
            insert_neuron_directed!(som, new_pos, weights)

            # Reset parent error after growing
            som.topology.neurons[pos].error = zero(T)

            n_added += 1
        end
    end

    return n_added
end

# ============================================================================
# Pruning
# ============================================================================

"""
    find_dead_neurons(topo; min_hits=1) -> Vector{Tuple{Int,Int}}

Find neurons that have been activated fewer than min_hits times.
"""
function find_dead_neurons(topo::SOMTopology; min_hits::Int=1)
    dead = Tuple{Int,Int}[]

    for (pos, neuron) in topo.neurons
        if neuron.hits < min_hits
            push!(dead, pos)
        end
    end

    return dead
end

find_dead_neurons(som::DBGSOM; min_hits::Int=1) =
    find_dead_neurons(som.topology; min_hits=min_hits)

"""
    prune_dead_neurons!(som; min_hits=1, min_neurons=4) -> Int

Remove neurons that have been activated fewer than min_hits times.
Keeps at least min_neurons in the map.

# Returns
Number of neurons removed
"""
function prune_dead_neurons!(som::DBGSOM; min_hits::Int=1, min_neurons::Int=4)
    dead = find_dead_neurons(som; min_hits=min_hits)
    n_removed = 0

    for pos in dead
        if n_neurons(som) > min_neurons
            remove_neuron!(som, pos)
            n_removed += 1
        end
    end

    return n_removed
end

File: R/growth.R
# growth.R - Growth, error distribution, and pruning for DBGSOM
#
# All functions that modify neurons RETURN the modified list.
# R lists have copy-on-modify semantics, so callers must reassign.

# ============================================================================
# Error Accumulation
# ============================================================================

#' Reset errors and hit counts for all neurons
#' @param neurons named list of neurons
#' @return modified neurons list
reset_errors <- function(neurons) {
  for (key in names(neurons)) {
    neurons[[key]]$error <- 0.0
    neurons[[key]]$hits <- 0L
  }
  neurons
}

#' Accumulate quantization errors for each neuron based on BMU assignments
#' @param neurons named list of neurons
#' @param X matrix (input_dim x n_samples)
#' @param bmu_keys character vector of BMU position keys
#' @param miss_info missingness info list or NULL
#' @return modified neurons list
accumulate_errors <- function(neurons, X, bmu_keys, miss_info = NULL) {
  n_samples <- ncol(X)

  for (i in seq_len(n_samples)) {
    key <- bmu_keys[i]
    w <- neurons[[key]]$weights
    x <- X[, i]

    if (!is.null(miss_info) && any(is.na(x))) {
      dist_sq <- euclidean_sq_na_weighted(w, x, miss_info)
    } else if (any(is.na(x))) {
      dist_sq <- euclidean_sq_na(w, x)
    } else {
      dist_sq <- euclidean_sq(w, x)
    }

    d <- sqrt(dist_sq)
    if (is.finite(d)) {
      neurons[[key]]$error <- neurons[[key]]$error + d
      neurons[[key]]$hits <- neurons[[key]]$hits + 1L
    }
  }
  neurons
}

# ============================================================================
# Error Distribution
# ============================================================================

#' Distribute errors from non-boundary to boundary neurons
#'
#' For non-boundary neuron with E >= GT:
#' - Each boundary neighbor gets error/12
#' - The non-boundary neuron retains error/2
#'
#' @param neurons named list of neurons
#' @param gt numeric growth threshold
#' @return modified neurons list
distribute_errors <- function(neurons, gt) {
  keys <- names(neurons)
  for (key in keys) {
    err <- neurons[[key]]$error
    pos <- key_to_pos(key)

    if (err >= gt && !is_boundary(neurons, pos)) {
      nbrs <- get_neighbors_hex(pos)
      boundary_nbrs <- character(0)
      for (nbr in nbrs) {
        nbr_key <- pos_to_key(nbr)
        if (!is.null(neurons[[nbr_key]]) && is_boundary(neurons, nbr)) {
          boundary_nbrs <- c(boundary_nbrs, nbr_key)
        }
      }

      if (length(boundary_nbrs) > 0) {
        error_share <- err / 12.0
        for (nbr_key in boundary_nbrs) {
          neurons[[nbr_key]]$error <- neurons[[nbr_key]]$error + error_share
        }
        neurons[[key]]$error <- err / 2.0
      }
    }
  }
  neurons
}

# ============================================================================
# Growth Candidates
# ============================================================================

#' Find boundary neurons with error exceeding growth threshold, sorted by error
#' @param neurons named list of neurons
#' @param gt numeric growth threshold
#' @return character vector of position keys (highest error first)
find_growth_candidates <- function(neurons, gt) {
  candidates <- character(0)
  errors <- numeric(0)

  for (key in names(neurons)) {
    pos <- key_to_pos(key)
    if (neurons[[key]]$error > gt && is_boundary(neurons, pos)) {
      candidates <- c(candidates, key)
      errors <- c(errors, neurons[[key]]$error)
    }
  }

  if (length(candidates) == 0) return(character(0))
  candidates[order(errors, decreasing = TRUE)]
}

# ============================================================================
# Directed Growth (Vasighi & Amini 2017)
# ============================================================================

#' Select the best growth direction from a boundary neuron
#'
#' Considers 1st and 2nd-order neighbors to bias growth toward high-error regions.
#'
#' @param neurons named list of neurons
#' @param boundary_key character position key of boundary neuron
#' @return list(pos_key, weights) or NULL if no growth possible
select_growth_direction <- function(neurons, boundary_key) {
  pos <- key_to_pos(boundary_key)
  parent <- neurons[[boundary_key]]
  allowed <- get_empty_neighbors(neurons, pos)

  if (length(allowed) == 0) return(NULL)

  # 1st-order occupied neighbors
  nbr1 <- get_neighbors_at_distance(neurons, pos, 1L)
  # 2nd-order occupied neighbors
  nbr2 <- get_neighbors_at_distance(neurons, pos, 2L)

  # Sort nbr1 by error (highest first)
  if (length(nbr1) > 1) {
    errs <- vapply(nbr1, function(p) neurons[[pos_to_key(p)]]$error, numeric(1))
    nbr1 <- nbr1[order(errs, decreasing = TRUE)]
  }
  if (length(nbr2) > 1) {
    errs <- vapply(nbr2, function(p) neurons[[pos_to_key(p)]]$error, numeric(1))
    nbr2 <- nbr2[order(errs, decreasing = TRUE)]
  }

  n_nbr1 <- length(nbr1)

  if (n_nbr1 == 0) {
    new_pos <- allowed[[1]]
    weights <- parent$weights + rnorm(length(parent$weights)) * 0.01
    return(list(pos_key = pos_to_key(new_pos), weights = weights))
  }

  if (n_nbr1 <= 2) {
    return(grow_few_neighbors(neurons, pos, parent, allowed, nbr1, nbr2))
  }

  if (n_nbr1 <= 4) {
    return(grow_medium_neighbors(neurons, pos, parent, allowed, nbr1))
  }

  return(grow_many_neighbors(neurons, pos, parent, allowed, nbr1))
}

#' Growth when 1-2 direct neighbors
grow_few_neighbors <- function(neurons, pos, parent, allowed, nbr1, nbr2) {
  sel_nbr2 <- list()
  for (p in nbr2) {
    for (a in allowed) {
      if (grid_distance_hex(p, a) == 1L) {
        sel_nbr2 <- c(sel_nbr2, list(p))
        break
      }
    }
  }

  if (length(sel_nbr2) > 0) {
    if (length(sel_nbr2) > 1) {
      errs <- vapply(sel_nbr2, function(p) neurons[[pos_to_key(p)]]$error, numeric(1))
      sel_nbr2 <- sel_nbr2[order(errs, decreasing = TRUE)]
    }
    best_nbr2_pos <- sel_nbr2[[1]]
    best_nbr2 <- neurons[[pos_to_key(best_nbr2_pos)]]
    dists <- vapply(allowed, function(a) grid_distance_hex(a, best_nbr2_pos), integer(1))
    new_pos <- allowed[[which.min(dists)]]
    weights <- (parent$weights + best_nbr2$weights) / 2.0
  } else {
    nbr1_pos <- nbr1[[1]]
    nbr <- neurons[[pos_to_key(nbr1_pos)]]
    dists <- vapply(allowed, function(a) grid_distance_hex(a, nbr1_pos), integer(1))
    new_pos <- allowed[[which.max(dists)]]
    weights <- 2.0 * parent$weights - nbr$weights
  }

  list(pos_key = pos_to_key(new_pos), weights = weights)
}

#' Growth when 3-4 direct neighbors
grow_medium_neighbors <- function(neurons, pos, parent, allowed, nbr1) {
  max_err_nbr_pos <- nbr1[[1]]
  min_err_nbr <- neurons[[pos_to_key(nbr1[[length(nbr1)]])]]

  dists <- vapply(allowed, function(a) grid_distance_hex(a, max_err_nbr_pos), integer(1))
  new_pos <- allowed[[which.min(dists)]]
  weights <- 2.0 * parent$weights - min_err_nbr$weights

  list(pos_key = pos_to_key(new_pos), weights = weights)
}

#' Growth when 5 direct neighbors
grow_many_neighbors <- function(neurons, pos, parent, allowed, nbr1) {
  new_pos <- allowed[[1]]

  opposite <- list()
  for (n in nbr1) {
    if (grid_distance_hex(n, new_pos) == 2L) {
      opposite <- c(opposite, list(n))
    }
  }

  if (length(opposite) > 0) {
    opp_nbr <- neurons[[pos_to_key(opposite[[1]])]]
    weights <- 2.0 * parent$weights - opp_nbr$weights
  } else {
    avg_w <- Reduce("+", lapply(nbr1, function(n) neurons[[pos_to_key(n)]]$weights)) / length(nbr1)
    weights <- 2.0 * parent$weights - avg_w
  }

  list(pos_key = pos_to_key(new_pos), weights = weights)
}

# ============================================================================
# Growth Step
# ============================================================================

#' Grow neurons from high-error boundary neurons using directed growth
#'
#' @param neurons named list of neurons
#' @param gt numeric growth threshold
#' @param max_neurons integer maximum neurons allowed
#' @return list with neurons (modified) and n_added (integer)
grow <- function(neurons, gt, max_neurons) {
  candidates <- find_growth_candidates(neurons, gt)
  n_added <- 0L

  for (key in candidates) {
    if (length(neurons) >= max_neurons) break

    result <- select_growth_direction(neurons, key)
    if (!is.null(result)) {
      neurons[[result$pos_key]] <- new_neuron(result$weights)
      neurons[[key]]$error <- 0.0
      n_added <- n_added + 1L
    }
  }

  list(neurons = neurons, n_added = n_added)
}

# ============================================================================
# Pruning
# ============================================================================

#' Find dead neurons (hit count < min_hits)
#' @param neurons named list of neurons
#' @param min_hits integer minimum hit threshold
#' @return character vector of dead neuron keys
find_dead_neurons <- function(neurons, min_hits = 1L) {
  dead <- character(0)
  for (key in names(neurons)) {
    if (neurons[[key]]$hits < min_hits) {
      dead <- c(dead, key)
    }
  }
  dead
}

#' Remove dead neurons, keeping at least min_neurons
#' @param neurons named list of neurons
#' @param min_hits integer minimum hit threshold
#' @param min_neurons integer minimum neurons to keep
#' @return list with neurons (modified) and n_removed (integer)
prune_dead_neurons <- function(neurons, min_hits = 1L, min_neurons = 4L) {
  dead <- find_dead_neurons(neurons, min_hits)
  n_removed <- 0L

  for (key in dead) {
    if (length(neurons) > min_neurons) {
      neurons[[key]] <- NULL
      n_removed <- n_removed + 1L
    }
  }

  list(neurons = neurons, n_removed = n_removed)
}
